### ISSUE TITLE
Add Pre-compare Script Option for DACPAC Publishing

### ISSUE THREAD

**Issue by dzsquared** [microsoft/DacFx#482]
https://developercommunity.visualstudio.com/t/Add-Pre-compare-Script-Option-for-DACPAC/10594162

Bringing this item over from VS Developer Community since the core implementation would be in Microsoft.Build.Sql and DacFx.

> Often times there are updates and revisions that need to be made to a database that cannot be automatically handled by a deployment script generated by a DACPAC, such as adding new non-nullable columns to a table or data migrations that must happen because a column or table is being dropped.

> The first instinct is to add a pre-deployment script with the code necessary to execute changes so that the DACPAC will publish correctly, but this fails because the pre-deployment script is run AFTER the comparison takes place to generate the deployment script. This has been a known source of confusion for at least 15 years (https://learn.microsoft.com/en-us/archive/blogs/gertd/pre-deployment-scripts) and there are still questions about it appearing on places like Stack Overflow and even issues being reported in the community feedback (https://developercommunity.visualstudio.com/t/SQL-database-project-pre-deployment-scri/224190?ref=native&refTime=1708466421969&refUserId=fd449ac5-d5e8-690b-8ba8-adeb29ac3e19).

> All of the workarounds require manually writing a change script to run before a DACPAC is published, which is fine, but then involve a variety of approaches to manage running the script outside of Visual Studio or the normal publishing process, or splitting known changes into separate source code updates that are checked in and published in stages, and so on. While these are all solutions that technically work, they are arguably confusing and convoluted.

> Would it be possible to add an additional feature to run a “pre-compare” script during publishing? This way users could include code for any changes they want to take place before a DACPAC performs a comparison. This would also enable developers to include this script as part of a .sqlproj, and have it automatically be handled as part of the normal publish process.

> The expected workflow would be:

> Run pre-compare script
> Compare and generate deployment script
> Run pre-deployment script
> Run deployment script
> Run post-deployment script

**Comment by tony-donley** [comment]
Thanks Drew! :-)

> The expected workflow would be:
> 
> Run pre-compare script
> Compare and generate deployment script
> Run pre-deployment script
> Run deployment script
> Run post-deployment script

It would also be nice to flag that 'pre-compare' script as '`One Time'` or '`Every Time`'.

'`One Time`' would be used for most advanced schema changes and data transformation, where the script should only execute once per target database and then be archived, not execute at every deployment.  Would that require something like a "schema version" mark in the target database to know if that script has executed or not?  Or would we just have to write our scripts to be stateless by reading the schema metadata and react conditionally?

'`Every Time` might do things like:
- set a maintenance flag earlier than the pre-deploy
- make temporary permission changes that are later reverted in the generated script or in post-deploy

**Comment by biltongza** [comment]
Chiming in here with my two cents, I love DACPACs for the mental model of "this is what my database should look like, make it happen" but this is an area where the experience could be better. As mentioned, sometimes there are things that I simply don't want the DACPAC to do, either because it can't (the non nullable column is an example) but also sometimes because the approach it takes to a change is not ideal and I want more control over it.

The other part of the experience that is not ideal, is that there is sometimes data that I want source controlled. For example static/seed data.

For this reason, I've landed up using something like DbUp twice: once as a pre-deployment step before the DACPAC, and once as a post-deployment step after the DACPAC. I use the pre-deployment for any changes like adding non nullable columns or anything else I don't want the DACPAC to change, and then the post-deployment for things like data.

I pick DbUp because it makes the experience much smoother, I don't need to focus as much on writing my scripts in an idempotent way because it uses a journaling table to know what has and hasn't been run yet.

It would be lovely if the DACPAC could handle this kind of scenario.

**Comment by ErikEJ** [comment]
Constantly seeing the need fir this, a simple option to simply execute a script would be so great

**Comment by dzsquared** [comment]
Apologies for the confusion @ErikEJ  - that roadmap item is calling out that for the script to get bundled into the dacpac for execution during deployment, Microsoft.Build.Sql would have to incorporate a new item. I would have better phrased it as "Pre-plan script support in DacFx previews (including Microsoft.Build.Sql and SqlPackage)"

The concept is still that it would impact deployment in this manner:

<img width="1564" height="874" alt="Image" src="https://github.com/user-attachments/assets/835d401a-0b68-4bc0-a6cb-a310bbfe693d" />

**Comment by dzsquared** [comment]
# Request for comments

We are working on adding a **pre-plan script** to SQL projects: a third script type, alongside [pre-deployment and post-deployment](https://learn.microsoft.com/en-us/sql/tools/sql-database-projects/concepts/pre-post-deployment-scripts?pivots=sq1-command-line) scripts, that runs before DacFx reverse-engineers the target database and compares it with the source model. The script would be included in the dacpac and could prepare the target for comparison, enabling scenarios such as data migrations required before making a column non-nullable or  before DacFx creates the deployment plan. Similar to pre/post-deployment scripts, a project can contain only one pre-plan script, with `:r` SQLCMD syntax available to compose it from multiple files.

```xml
<ItemGroup>
  <PrePlan Include="pre-plan.sql" />
</ItemGroup>
```

Before we finalize the design, we want to get feedback on the proposed behavior and limitations of pre-plan scripts. Any feedback on the proposed behavior, limitations, and scenarios is welcome. Please comment on this issue and we'll continue to provide updates as we release previews of the Microsoft.Build.Sql SDK and SqlPackage CLI that include pre-plan script support.

**Limitations of pre-plan scripts**:
- The T-SQL syntax in a pre-plan script must be able to be run inside a transaction. Operations like fulltext catalog/index operations, resumable index operations, and backup/restore are not supported. Build-time checks will make best effort to flag unsupported operations.
- Similar to pre/post-deployment scripts, pre-plan scripts should be authored with idempotent guardrails to avoid errors if the script is run multiple times. Upcoming functionality will surface dacpac versioning in a way that makes it easier to ensure portions of a script are only run once throughout a database's lifecycle.

For **Publish** actions, the proposal is to run the pre-plan script in a transaction before model comparison. If the script or deployment plan generation fails, the pre-plan script transaction would be rolled back; successfully beginning the deployment would commit the pre-plan script. Even if the publish property `IncludeTransactionalScripts` has been set to true, the pre-plan script would still be handled as a separate transaction that is committed once the deployment plan is successfully generated.

For **Script** actions, the proposal is to require the explicit choice of one of two modes for how the pre-plan script is handled. The script action is expected to be *read-only*, so the presence of a pre-plan script requires a decision to ensure that the executor is aware of the tradeoffs. The two modes are:
- /p:PrePlanScript=RunAndRollback — The pre-plan script is executed in a transaction against the target, the deployment plan is generated against the modified state, and then the transaction is rolled back. The pre-plan script is included in the script output returned to the user.
- /p:PrePlanScript=RunOnce — The pre-plan script is executed when generating the deployment plan, and if the plan generates successfully the pre-plan script transaction is committed. The pre-plan script is not included in the script output, since it has already been applied to the target.

 **Is requiring an explicit choice the right default for `Script` actions?** We especially want feedback on whether temporary execution under `RunAndRollback` meets your read-only expectations, whether you need a mode that never touches the target even if its plan may be inaccurate, and whether you would use `RunOnce` despite Script then making a persistent change.
- Does requiring the property selection for `/p:PrePlanScript` between "RunOnce" and "RunAndRollback" and having no default option make using pre-plan scripts with the Script action for deployments seem safer, less like running with scissors?
- Is the "RunOnce" property option a mode that you might use in the environments you deploy to?

**Comment by ErikEJ** [comment]
@dzsquared Awesome! 

"Is requiring an explicit choice the right default for Script actions?" 

not sure what you mean by this... so /a:script will not run if the .dacpac contains a preplan script unless you set one of the two options?

In that case RunAndRollback sounds like a good approach.

**Comment by dzsquared** [comment]
@ErikEJ - you're correct, the proposal is that /a:script will error/not run if there's a pre-plan script in the dacpac and the property /p:preplanscript hasn't been set. The intention is to keep anyone from being caught off guard by a script (even though they're rolled back). What we're very interested in is:
- does requiring option selection make pre-plan scripts and the /a:script action make them seem less like running with scissors (more safe)
- is the "runonce" option seem like something you might ever use, based on the environments you're deploying to

I'll edit the final question segment for clarity.

**Comment by ErikEJ** [comment]
@dzsquared I never actually use /a:script, but think that forcing users to set the option makes a lot of sense. On the other hand, if using scripts, I would never use RunOnce, as I assume this purpose of using scripts will then delay the actual publish process.

(My scripts usually contain destructive actions like column and table drops, and the occasional rename, all pretty breaking)

**Comment by CHerting-NI** [comment]
We implemented our own "pre-pre-deployment" script. But a solution from the tool would be very helpful.

**Comment by dzsquared** [comment]
@CHerting-NI - do you use your pre-pre-deployment scripts with the "script" action? do you have specific behavior expectations of it?

**Comment by CHerting-NI** [comment]
@dzsquared : Yes, with the script action. My expectation is that I have a script that runs automatically before database compare.

**Comment by ErikEJ** [comment]
@dzsquared an example of my current pipeline flow:

```yaml
                      - task: SqlAzureDacpacDeployment@1
                        displayName: Deploy DataLoss script using access token
                        inputs:
                          azureSubscription: ${{ parameters.serviceConnection }}
                          AuthenticationType: 'connectionString'
                          ConnectionString: 'Data Source=$(sqlServerFqdn);Initial Catalog=$(sqlDatabaseName);Encrypt=true;Connect Timeout=60'
                          deployType: 'SqlTask'
                          sqlFile: '$(Pipeline.Workspace)/drop/DataLossScript.sql'
                          SqlAdditionalArguments: '-AccessToken $(sqlAccessToken)'

                      - task: SqlAzureDacpacDeployment@1
                        displayName: Deploy .dacpac using access token
                        inputs:
                          azureSubscription: ${{ parameters.serviceConnection }}
                          AuthenticationType: 'connectionString'
                          ConnectionString: 'Data Source=$(sqlServerFqdn);Initial Catalog=$(sqlDatabaseName);Encrypt=true;Connect Timeout=60'
                          deployType: 'DacpacTask'
                          DeploymentAction: 'Publish'
                          DacpacFile: '$(Pipeline.Workspace)/drop/Database.Build.dacpac'
                          AdditionalArguments: '/AccessToken:$(sqlAccessToken)'
```

**Comment by DanielLoth** [comment]
Hi @dzsquared,

Does it have to nest the pre-plan capability within a transaction?
Would it instead be possible for that to be configurable and left to the discretion of users of the tool?

The reason I say that is, for better or worse, sqlpackage often takes the 'easy' way out when it comes to doing certain things.

For example, let's say I want to drop an index.

The sqlpackage tool will produce something like this:
```
drop index IDX1 on dbo.T1;
go
```

And that's fine on a quiet database.

On a busy database, attempting to drop the index in this manner has the propensity to cause blocking - even on Enterprise Edition.

Unlike the `create index` statement, there is no `online = on` option when dropping an index. There's also no notion of dropping an index at a lower priority built into the T-SQL syntax.

The drop index statement queues up for its `Sch-M` lock, and suddenly regular OLTP queries against the table have to wait for the `Sch-S` lock which can only be granted after the drop index statement gets to run.

So to do it safely, we'd actually need to run something like the following (and this is more or less how I routinely do this kind of thing in my real-world work):
```
set nocount, xact_abort on;
set deadlock_priority low;
set lock_timeout 500;

declare @Delay datetime = dateadd(millisecond, 500, convert(datetime, 0x0));
declare @Deadline datetime = dateadd(minute, 3, getutcdate());
declare @CanRetry bit = 1;

while @CanRetry = 1
begin
	begin try
		exec('drop index if exists IDX1 on dbo.T1;');
		break;
	end try
	begin catch
		if @@trancount != 0 rollback;

		set @CanRetry = iif(
			getutcdate() <= @Deadline and
			error_number() in (1205, 1222),
			1,
			0
		);

		if @CanRetry = 0 throw;

		waitfor delay @Delay;
	end catch
end
```

Something like the above snippet would be brilliant to incorporate as-is in the pre-plan stage you've proposed. We'd currently run such a change manually and then incorporate the change into the version-controlled code that is used to build the DACPAC (which we refer to as 'washing back a change').

But to do that, or any other task that has the potential to cause contention in a contention-sensitive database, I need to have ownership over the lifecycle of a transaction.

Alternatively, it'd be good if the deployment contributors could more readily morph the T-SQL generated by certain deployment steps.
A lot of deployment steps are `sealed`, and immutable, classes that you can't instantiate in your deployment contributor code. It's challenging to use a deployment plan modifier for the purposes of making the main phase of a deployment 'safe' in high contention servers where the goal is near-zero downtime.

I've always gotten the impression that deployment plan modifiers are more intended to _subtract_ steps rather than substitute steps with more robust versions, or to just adorn existing steps with extra options (e.g.: adding `online = on` or `maxdop 1` to a create index statement).

**Comment by Mnior** [comment]
> Does it have to nest the pre-plan capability within a transaction?
> Would it instead be possible for that to be configurable and left to the discretion of users of the tool?

Your question is alarming. Because there is already a setting:
[`/p:IncludeTransactionalScripts=(BOOLEAN)`](https://learn.microsoft.com/en-us/dotnet/api/microsoft.sqlserver.dac.dacdeployoptions.includetransactionalscripts)
_`Publish.xml`_
``` xml
<Project ...>
  <PropertyGroup>
    <IncludeTransactionalScripts>True/False</IncludeTransactionalScripts>
  <PropertyGroup>
</Project>
```
[IncludeTransactionalScripts](https://learn.microsoft.com/en-us/dotnet/api/microsoft.sqlserver.dac.dacdeployoptions.includetransactionalscripts)

And I assumed that the specified setting extends to this block, and works uniformly in meaning.
Moreover, even the inclusion of this parameter does not guarantee that the base remains unchanged, since it is necessary that scripts are made in a special way.

With a mandatory transaction, this functionality moves from the category of a dubious thing to a completely useless idea.
We have used a separate PreDepoy step before Publish and will continue to do so, regardless of the availability of this functionality.

**Comment by DanielLoth** [comment]
These are excerpts from @dzsquared's message above:

> The T-SQL syntax in a pre-plan script must be able to be run inside a transaction

> Even if the publish property IncludeTransactionalScripts has been set to true, the pre-plan script would still be handled as a separate transaction that is committed once the deployment plan is successfully generated.

As I think a little more about our own workflow, after a closer read of the above messages, I'm not quite sure how it'd work out for us.

We use the `Script` and `DeployReport` actions before using the `Publish` action. The script and deployment report are then made available for a reviewer to review (or, alternatively, programmatically confirmed safe via analysis of the report XML), and the deployment pipeline is suspended at that point until approval is granted.

The proposed approach to nest the pre-plan phase in a single transaction - and notably a single long-running transaction if the deployment planning phase takes multiple seconds - probably means the feature is dead on arrival for us.

One thing to keep in mind, too, is that people using deployment contributors for certain use cases will necessarily have a longer, slower, deployment planning phase.

One of our specific deployment contributor use cases involves the following:
- Connect to the database that the deployment is being performed against.
- Query the metadata DMVs pertaining to security principals and their existing permissions.
- Determine whether or not any permissions grants that are in the deployment plan are redundant (i.e., granting a permission that the target security principal already holds).
- Remove those permissions grants from the deployment plan.

Admittedly this use case _shouldn't_ be required. However, for whatever reason, we have found that sqlpackage sometimes errs and determines that a permission grant is required when in fact it is not. In a less busy database, it wouldn't matter - the grant statements are idempotent and quite harmless. But because our database is quite busy, we very frequently get deadlocks due to permission grants (and removing the redundant ones reduces the probability of this occurring).

This whole deployment contributor phase would now all be happening while the singular pre-plan transaction is open and pending commit/rollback. And not once, but three times given that we use `Script`, `DeployReport`, then have a human review the generated deployment report (or programmatically approve it if we deem the report to indicate that the deployment is 'safe'), and only then conclude with the `Publish` action.

**Comment by DanielLoth** [comment]
At the risk of a little sidebar conversation, it's worth noting here that the pre-plan mechanism - at least in the use cases I've mentioned - is not fundamentally required.

If there was a richer experience during the deployment plan phase (e.g.: the ability to intercept, and override, the T-SQL that will run for a certain deployment steps - rather than just the ability to adorn the statements that have already been decided upon with certain options) then the outcome would be the same: A reduction in the number of workarounds required when the desired state promise breaks down.

That is, after all, what the pre-plan phase is for as best I can tell - working around a breakdown in the promise of delivery on desired state.

---

In the case of a permission grant, for example, we could trivially augment the code from something like this:
```
grant select on dbo.T1 to U1;
```

To something like this:
```
if not exists (
	select 1
	from sys.database_permissions p
	where
		p.grantee_principal_id = user_id('U1') and
		p.major_id = object_id('dbo.T1')
)
begin
	grant select on dbo.T1 to U1;
end
```

Then we'd have accomplished our goal within the desired state model (without needing a deployment contributor).

---

Same concept with the drop index statement. If we could better augment the code that gets created when dropping an element (e.g.: drop index) then there'd be no need to do it during a pre-plan phase. The same can be said for creating an element where that element can't be done online (e.g.: creating a new foreign key).

The creating a new foreign key example might be a little confusing. Fundamentally though, you can minimise locking and blocking during foreign key constraint creation by creating it with nocheck.
But what you can't do within an `alter table ... add foreign key` statement is run it with `online = on`, or at a low priority so that it's wait for the initial `Sch-M` lock doesn't become a leading blocker for a lot of subsequent OLTP queries that want the `Sch-S` lock on either the parent or referenced table.

The same applies to check constraints. You can add the check constraint with nocheck to hasten its execution (by skipping the checking of existing rows, which necessarily blocks the table for the duration since the creation of a check constraint is not accomplishable online), but you can't fundamentally resolve the window of time where the `alter table ... add check` statement is blocked by OLTP queries using the table, and then itself becomes a lead blocker for more OLTP queries that are awaiting its conclusion.

---

EDIT: Fixed example code for enhanced permission grant that checks the grant is actually required before attempting to run the grant statement.

---

EDIT 2: Just to add: Yet another alternative is that Microsoft update the default code that gets produced to use more complicated/defensive patterns by default (or if you specify some parameter).
If the default `grant ...` statement batch that gets produced is defensive in nature by default, then suddenly it's no longer a concern.
Same applies to a whole array of changes that one might deploy out in a busy database.

This too would bridge the gap that results in a publish using sqlpackage being deemed 'unsafe', and thus remove the need completely to precede the release with a manual change (as is our situation) or some kind of pre-plan step.

**Comment by DanielLoth** [comment]
Oh, and I just want to clarify one more thing... Because I recognise that it'll cause confusion.

I know that you can replace any step - such as a create element step - in the deployment plan. You can use the regular `DeploymentScriptDomStep` from memory which is the base class for all of the much richer classes such as the following:
- https://learn.microsoft.com/en-us/dotnet/api/microsoft.sqlserver.dac.deployment.dropelementstep?view=sql-dacfx-162
- https://learn.microsoft.com/en-us/dotnet/api/microsoft.sqlserver.dac.deployment.createelementstep?view=sql-dacfx-162
- https://learn.microsoft.com/en-us/dotnet/api/microsoft.sqlserver.dac.deployment.sqltablemigrationstep?view=sql-dacfx-162

But as soon as you remove one of these _rich_ steps from the deployment plan and replace them with just an instance of `DeploymentScriptDomStep`, you no longer see the things you replaced within the deployment report.
